src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,987 @@ mod v1 {
+                                                      cf_file.path,
+                                                      cf_file.cf,
+                                                      checksum,
+                                                      cf_file.checksum)));
+                }
+            }
+            let mut total_size = 0;
+            for cf_file in &mut self.cf_files {
+                try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                total_size += cf_file.size;
+            }
+            // write meta file
+            let mut v = vec![];
+            try!(self.meta_file.data.write_to_vec(&mut v));
+            try!(self.meta_file.file.take().unwrap().write_all(&v[..]));
+            try!(fs::rename(&self.meta_file.tmp_path, &self.meta_file.path));
+            let mut size_track = self.size_track.wl();
+            *size_track = size_track.saturating_add(total_size);
+            Ok(())
+        }
+
+        fn apply(&mut self, context: &mut ApplyContext) -> Result<()> {
+            for (cf, path, size) in box_try!(self.list_cf_path_sizes()) {
+                if size == 0 {
+                    // Skip ingesting empty sst file since that would cause an error
+                    // "Corruption: file is too short to be an sstable"
+                    continue;
+                }
+                try!(check_abort(&context.abort));
+
+                let cf_handle = box_try!(rocksdb::get_cf_handle(&context.db, &cf));
+                let ingest_opt = IngestExternalFileOptions::new().move_files(true);
+                box_try!(context.db.ingest_external_file_cf(cf_handle, &ingest_opt, &[&path]));
+                context.snapshot_size += size as usize;
+            }
+            Ok(())
+        }
+    }
+
+    impl Read for Snap {
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            if buf.len() == 0 {
+                return Ok(0);
+            }
+            while self.cf_index < self.cf_files.len() {
+                match self.cf_files[self.cf_index].file.as_mut().unwrap().read(buf) {
+                    Ok(0) => {
+                        // EOF. Switch to next file.
+                        self.cf_index += 1;
+                    }
+                    Ok(n) => {
+                        return Ok(n);
+                    }
+                    e => return e,
+                }
+            }
+            Ok(0)
+        }
+    }
+
+    impl Write for Snap {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            if buf.len() == 0 {
+                return Ok(0);
+            }
+            let mut next_buf = buf;
+            while self.cf_index < self.cf_files.len() {
+                let cf_file = &mut self.cf_files[self.cf_index];
+                let left = (cf_file.size - cf_file.written_size) as usize;
+                if left == 0 {
+                    self.cf_index += 1;
```

Comment:
The order is assigned in the snapshot meta. Since it uses length specified protocol and validates the checksums, it's unnecessary to add delimiters.